test/index.js

```diff
@@ -1589,6 +1589,33 @@ describe('def', function() {
                    '  - (Either ??? ???)\n'));
   });
 
+  it('supports values of "foreign" types', function() {
+    //  id :: a -> a
+    var id = def('id', {}, [a, a], R.identity);
+
+    eq(id({'@@type': 'my-package/Foreign'}), {'@@type': 'my-package/Foreign'});
+
+    throws(function() { id([{'@@type': 'my-package/Foo'},
+                            {'@@type': 'my-package/Bar'}]); },
+           errorEq(TypeError,
+                   '‘id’ received [{"@@type": "my-package/Foo"}, ' +
+                   '{"@@type": "my-package/Bar"}] as its first argument, ' +
```

Comment:
This doesn't seem to be the best possible error message for the user. From their point of view, both `my-package/Foo` and `my-package/Bar` are valid types, but they don't see these types in the list below. I understand that `Array ???` captures the specific `TypeError` here, but it would be quite difficult to discern for most people. Maybe we should have a specific error message for an `Array` with no shared consistent type for all values?